sw/device/lib/dif/dif_pinmux.h

```diff
@@ -0,0 +1,654 @@
+typedef enum dif_pinmux_wakeup_mode_timed {
+  /**
+   * Trigger a wakeup request when pin is driven HIGH for a certain amount of
+   * always-on clock cycles as configured in
+   * `dif_pinmux_wakeup_timed_config_t`, `counter_threshold` field.
+   */
+  kDifPinmuxWakeupModeTimedHigh = 0,
+  /**
+   * Trigger a wakeup request when pin is driven LOW for a certain amount of
+   * always-on clock cycles as configured in
+   * `dif_pinmux_wakeup_timed_config_t`, `counter_threshold` field.
+   */
+  kDifPinmuxWakeupModeTimedLow,
+} dif_pinmux_wakeup_mode_timed_t;
+
+/**
+ * A Pin Multiplexer timed wake-up mode configuration.
+ */
+typedef struct dif_pinmux_wakeup_timed_config {
+  /**
+   * Wake-up detection mode.
+   */
+  dif_pinmux_wakeup_mode_timed_t mode;
+  /**
+   * Counter threshold for `kDifPinmuxWakeupModeTimedLow` and
+   * `kDifPinmuxWakeupModeTimedHigh` wake-up detector modes. The threshold is in
+   * terms of always-on clock cycles.
+   */
+  uint8_t counter_threshold;
+  /**
+   * Common configuration between different wake-up modes.
+   */
+  dif_pinmux_wakeup_config_t common;
+} dif_pinmux_wakeup_timed_config_t;
+
+/**
+ * The result of a Pin Multiplexer operation.
+ */
+typedef enum dif_pinmux_result {
+  /**
+   * Indicates that the operation succeeded.
+   */
+  kDifPinmuxOk = 0,
+  /**
+   * Indicates some unspecified failure.
+   */
+  kDifPinmuxError = 1,
+  /**
+   * Indicates that some parameter passed into a function failed a
+   * precondition.
+   *
+   * When this value is returned, no hardware operations occurred.
+   */
+  kDifPinmuxBadArg = 2,
+} dif_pinmux_result_t;
+
+/**
+ * Creates a new handle for a Pin Multiplexer.
+ *
+ * This function does not actuate the hardware.
+ *
+ * @param params Hardware instantiation parameters.
+ * @param[out] pinmux Out param for the initialized handle.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_pinmux_result_t dif_pinmux_init(dif_pinmux_params_t params,
+                                    dif_pinmux_t *pinmux);
+
+/**
+ * Locks out Pin Multiplexer functionality based on the `target` and `index`.
+ *
+ * This function allows for a fine grained locking of the Pin Multiplexer
+ * parts. `index` changes meaning dependent on the `target`. For example, when
+ * `target` is `kDifPinmuxLockTargetInsel`, `index` represents a peripheral
+ * input for which the MIO input select functionality should be locked.
+ *
+ * NOTE:
+ * Locking an already locked register will succeed with the `kDifPinmuxOk`
+ * return code.
+ *
+ * @param pinmux A Pin Multiplexer handle.
+ * @param index Exact meaning is determined by the `target`, however, it
+ *              represents one of the following: peripheral input index,
+ *              MIO pad index, DIO pad index or a wakeup detector index.
+ * @param target Target functionality to be locked.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_pinmux_result_t dif_pinmux_lock(dif_pinmux_t *pinmux,
```

Comment:
Here the `dif_pinmux_t *` is not `const` but in other places it is `const`. I don't think anything actually writes to this data structure apart from the `dif_pinmux_init` function, so maybe we should make it `const` everywhere?

Reply:
Good catch.

Done.